### PR TITLE
[train_text_to_image] allow using non-ema weights for training

### LINKED ISSUE #1153: Why does train_text_to_image.py perform so differently from the CompVis script?
I posted about this on the forum but didn't get any useful feedback - would love to hear from someone who knows the in and outs of the diffusers codebase!

https://discuss.huggingface.co/t/discrepancies-between-compvis-and-diffuser-fine-tuning/25556

To summarize the post: the `train_text_to_image.py` script and original CompVis repo perform very differently when fine-tuning on the same dataset with the same hyperparameters. I'm trying to reproduce the Lamda Labs Pokemon fine-tuning results and finding difficulty doing so (picture results in forum post).

I've been digging into the implementations and I'm not noticing any obvious differences in how the models are trained, losses are calculated, etc - so what explains the large behavioral discrepancies?

Would really appreciate any insight on what might be causing this.

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR allows using `non-ema` weights for training and `ema` weights for EMA updates to mimic the original training process.  For now, the workflow is as follows

- Each pre-trained SD checkpoint will have a branch called `non-ema`. 
- The script allows specifying this using the `--non_ema_revision` argument. If it's `None` it will default to using ema weights for both training and ema as is the case now.
- if `--non_ema_revision` is specified it will be used to load the `unet` for training and the ema (`main`) weights will be used for ema updates.

This approach of using branches is not the best solution but will be used until we have the `variations` feature in `diffusers`.

This PR also allows checkpointing of the ema model. Currently only the train `unet` is checkpointed.

Example command:

```bash
export MODEL_NAME="runwayml/stable-diffusion-v1-5"
export NOM_EMA_REVISION="non-ema"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"
export WANDB_PROJECT="stable-diffusion-pokemon"

accelerate launch --multi_gpu --gpu_ids="0,1" --mixed_precision="no" \
   ../diffusers/examples/text_to_image/train_text_to_image.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --non_ema_revision=$NOM_EMA_REVISION --use_ema \
  --dataset_name=$DATASET_NAME --caption_column="text" \
  --resolution=512 --center_crop --random_flip \
  --train_batch_size=16 --gradient_checkpointing \
  --max_train_steps=5000 --checkpointing_steps=1000 \
  --learning_rate=3e-05 \
  --lr_scheduler="constant" --lr_warmup_steps=0 \
```

Fixes #1153

### CODE REVIEW

File: examples/text_to_image/train_text_to_image.py [hunk 1]
```diff
@@ -172,6 +173,13 @@ def parse_args():
         "--use_8bit_adam", action="store_true", help="Whether or not to use 8-bit Adam from bitsandbytes."
     )
     parser.add_argument("--use_ema", action="store_true", help="Whether to use EMA model.")
+    parser.add_argument(
+        "--non_ema_revision",
+        type=str,
+        default=None,
+        required=False,
+        help="Revision of pretrained non-ema model identifier from huggingface.co/models.",
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:181]:
```suggestion
        help="Revision of pretrained non-ema model identifier. Must be a branch, tag or git identifier of the local or remote repository specified with --pretrained_model_name_or_path.",
```

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
             for p in self.shadow_params
         ]
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        """
+        Save the current parameters for restoring later.
+        Args:
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored. If `None`, the parameters of with which this
+                `ExponentialMovingAverage` was initialized will be used.
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:347]:
```suggestion
                temporarily stored.
```

Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:347]:
```suggestion
                temporarily stored.
```

File: examples/text_to_image/train_text_to_image.py [hunk 3]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
...
+        """
+        parameters = list(parameters)
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:349]: Very minor question, why do we need the conversion to `list` here?

File: examples/text_to_image/train_text_to_image.py [hunk 4]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
...
+        Restore the parameters stored with the `store` method.
+        Useful to validate the model with EMA parameters without affecting the
+        original optimization process. Store the parameters before the
+        `copy_to` method. After validation (or model saving), use this to
+        restore the former parameters.
+        Args:
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                updated with the stored parameters. If `None`, the
+                parameters with which this `ExponentialMovingAverage` was
+                initialized will be used.
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:363]: I don't think `None` is supported as described here.

```suggestion
                updated with the stored parameters.
```

File: examples/text_to_image/train_text_to_image.py [hunk 5]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
...
+
+        self.collected_params = state_dict["collected_params"]
+        if self.collected_params is not None:
+            assert isinstance(self.collected_params, list), "collected_params must be a list"
+            assert all(
+                isinstance(p, torch.Tensor) for p in self.collected_params
+            ), "collected_params must all be Tensors"
+            assert len(self.collected_params) == len(
+                self.shadow_params
+            ), "collected_params and shadow_params had different lengths"
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:412]:
```suggestion
            ), "collected_params and shadow_params must have the same length"
```

(same structure as previous assertions)

File: examples/text_to_image/train_text_to_image.py [hunk 6]
```diff
@@ -599,7 +724,6 @@ def collate_fn(examples):
             dirs = sorted(dirs, key=lambda x: int(x.split("-")[1]))
             path = dirs[-1]
         accelerator.print(f"Resuming from checkpoint {path}")
-        accelerator.load_state(os.path.join(args.output_dir, path))
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:602]: Why is `resume_from_checkpoint` disabled?

File: examples/text_to_image/train_text_to_image.py [hunk 7]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
...
+        #             )
+        #             if self.collected_params is not None:
+        #                 self.collected_params[i] = self.collected_params[i].to(
+        #                     device=p.device, dtype=p.dtype
+        #                 )
+        # else:
+        #     raise ValueError(
+        #         "Tried to `load_state_dict()` with the wrong number of "
+        #         "parameters in the saved state."
+        #     )
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:434]: Maybe remove this when we are done :)

File: examples/text_to_image/train_text_to_image.py [hunk 8]
```diff
@@ -324,6 +338,101 @@ def to(self, device=None, dtype=None) -> None:
...
+            raise RuntimeError("This ExponentialMovingAverage has no `store()`ed weights to `restore()`")
+        parameters = list(parameters)
+        for c_param, param in zip(self.collected_params, parameters):
+            param.data.copy_(c_param.data)
+
+        self.collected_params = None
+        torch.cuda.empty_cache()
+
+    def state_dict(self) -> dict:
+        r"""Returns the state of the ExponentialMovingAverage as a dict."""
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:375]: Suggestion: maybe comment somewhere that `state_dict`, `load_state_dict` are required by `accelerate` so that checkpointing works (as they are not used directly in the script).

**Review by pcuenca — COMMENTED** [review]
Looks great, thanks a lot!

However, I may not be fully understanding it yet. I know that `state_dict` and `load_state_dict` are used by `accelerate` during the checkpointing process, but I don't understand how `store` and `restore` are used. In addition, the line to resume from a checkpoint appears to have been removed, is resuming performed differently now?

File: examples/text_to_image/train_text_to_image.py (hunk 6 above)
Comment (patil-suraj) [reply examples/text_to_image/train_text_to_image.py:602]: Oops, this is removed by mistake.

**Comment by patil-suraj** [comment]
>  but I don't understand how store and restore are used.

The `store` and `restore` methods can be used for evaluation during training. For evaluation we want the EMA params, so we save the training params temporarily in EMAModel using the `store`, copy the ema params to the model, do the evaluation and then use the `restore` to restore the training parameters, back. But this is not currently used in the script, so will remove these methods.

  
> In addition, the line to resume from a checkpoint appears to have been removed, is resuming performed differently now?
My bad, removed it by mistake.

File: examples/text_to_image/train_text_to_image.py [hunk 9]
```diff
@@ -361,39 +435,44 @@ def main():
...
-    vae.requires_grad_(False)
-    text_encoder.requires_grad_(False)
-
     if args.gradient_checkpointing:
         unet.enable_gradient_checkpointing()
 
+    # Enable TF32 for faster training on Ampere GPUs,
+    # cf https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices
+    if args.allow_tf32:
+        torch.backends.cuda.matmul.allow_tf32 = True
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:474]: This gives ~1.3x speed-up on A100.

File: examples/text_to_image/train_text_to_image.py [hunk 10]
```diff
@@ -541,7 +614,8 @@ def collate_fn(examples):
     unet, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
         unet, optimizer, train_dataloader, lr_scheduler
     )
-    accelerator.register_for_checkpointing(lr_scheduler)
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:544]: It's not required to register `lr_scheduler` here, it's automatically checkpointed by `accelerate`. We only need to register custom objects.

File: examples/text_to_image/train_text_to_image.py [hunk 11]
```diff
@@ -482,9 +572,10 @@ def tokenize_captions(examples, is_train=True):
                 raise ValueError(
                     f"Caption column `{caption_column}` should contain either strings or lists of strings."
                 )
-        inputs = tokenizer(captions, max_length=tokenizer.model_max_length, padding="do_not_pad", truncation=True)
-        input_ids = inputs.input_ids
-        return input_ids
+        inputs = tokenizer(
+            captions, max_length=tokenizer.model_max_length, padding="max_length", truncation=True, return_tensors="pt"
+        )
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image.py:577]: always padding to `max_length` now to completely match with original implem.

File: examples/text_to_image/train_text_to_image.py (hunk 10 above)
Comment (pcuenca) [reply examples/text_to_image/train_text_to_image.py:544]: Interesting, thanks. [This documentation](https://huggingface.co/docs/accelerate/usage_guides/checkpoint) led me to believe we needed to register it, but in those examples the learning rate scheduler is not being passed to `prepare`.

**Review by pcuenca — APPROVED** [review]
Looks great!

File: examples/text_to_image/train_text_to_image.py (hunk 10 above)
Comment (patrickvonplaten) [reply examples/text_to_image/train_text_to_image.py:544]: Hmm should we maybe ask on the `accelerate` repo?

**Review by patrickvonplaten — APPROVED** [review]
Looks clean! I trust @pcuenca and @patil-suraj here :-)

File: examples/text_to_image/train_text_to_image.py (hunk 10 above)
Comment (patil-suraj) [reply examples/text_to_image/train_text_to_image.py:544]: I've verified this, all standard objects that we pass to `prepare` (like `nn.Module`, `DataLoader`, `Optimizer`, `Scheduler`) are automatically checkpointed by `accelerate`. We only need to register custom objects or models that we don't pass to `prepare`.
